enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,272 @@
+---
+title: advertise-version-details
+authors:
+  - "@dhellmann"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@spadgett, console team"
+  - "@wking, upgrades team"
+  - "@sdodson, upgrades team"
+  - "@LalatenduMohanty, upgrades team"
+  - "@soltysh, workloads team"
+approvers:
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2022-07-29
+last-updated: 2022-07-29
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/USHIFT-233
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1850656
+---
+
+# Advertise Version Details
+
+## Summary
+
+We have several use cases for allowing authenticated users to discover
+whether they are communicating with OCP or MicroShift, including the
+version of the product. We already have the ClusterVersion resource
+for OCP, but that requires special privileges to read and is not
+present in a MicroShift deployment. This enhancement describes a
+standard that both platforms can use to share platform and version
+information with API callers.
+
+## Motivation
+
+Users configuring applications to deploy on different versions of
+OpenShift may need to adjust their configuration to work differently
+on OCP and MicroShift. Ideally this is not needed, but in practice it
+is useful because of different APIs being available and different
```

Comment:
If you are wondering "is API X available?", one path is:

1. What product am I running?  MicroShift
2. What version am I running?  x.y.z
3. Ah, my lookup table suggests that MicroShift x.y.z includes API X.
4. Use or talk about API X.

Another is:

1. Attempt to poke API X.
2. API server responds: "I have no API X".
3. Do not use or talk about API X.

The second approach does not require keeping a lookup table about which products/versions have which APIs.  It also gracefully handles cases where API X _should_ be installed, but for some reason is not working at the moment.  The downside is that it's slightly more dynamic to check all the APIs you're curious about, but you're already dynamically polling product/version, so it's not a fundamental shift there.

Reply:
That's fair. However, not everything is discoverable by probing for an API. In practice we do see code looking at the version and doing different things. It might configure a Deployment differently (1 replica, smaller requests, etc.) for example.

It's also significantly easier in some contexts to have a single check for "is OpenShift" rather than probing each individual API. Even [our own products](https://github.com/stackrox/helm-charts/blob/main/3.71.0/central-services/templates/_openshift.tpl#L21) do this.

Reply:
I added a use case for this below.

Reply:
> It might configure a Deployment differently (1 replica, smaller requests, etc.) for example.

Do we need to expand what we're publishing here to include attributes beyond version?

Reply:
> > It might configure a Deployment differently (1 replica, smaller requests, etc.) for example.
> 
> Do we need to expand what we're publishing here to include attributes beyond version?

There may be other useful information we could export (topology is excluded here but could be used by an installer), but I think I'd rather limit this enhancement to just the version strings for now and deal with the other values in a separate enhancement when we have a firm use case. Unless you had something specific in mind that seems critical?

Reply:
yeah, I agree. I would perfer to reduce the scope to things which has a solid usecase now.

Reply:
Sorry to unresolve, but here's a use case (feel free to re-resolve and say "yeah, we can cover this later"):

We've talked about moving operator authors away from declaring constraints on cluster versions and toward declaring constraints on specific APIs or cluster properties. We've imagined that an OLM component could probe the cluster and receive a list of provided APIs (theoretically, this is already possible with the discovery API today), but also potentially arbitrary properties (e.g. "openshift.rhcosVersion" or "openshift.varaint"="hypershift|microshift|standard|okd|etc.").

Without fine-grained properties like this, operator authors are typically conservative with their maxOpenshiftVersion value. They only declare what they can actually test against, so they typically set maxOpenshiftVersion no further than the most recent version. This bleeds over to cluster admins, who are then blocked from upgrading to the next version of OpenShift, even if the operators they had installed actually _would_ be compatible. This then forces operator authors to publish "new" bundle versions in existing catalogs with an updated "maxOpenshiftVersion" value to unblock cluster admins and allow them to upgrade to the next OCP minor version.

If operator authors could declare more fine-grained constraints against cluster attributes, it would mean a significant reduction in toil for operator authors in that they would no longer have to produce new bundle versions just to declare support for the upcoming version of OCP.

Reply:
@joelanford That use case makes sense, but I think I would want OLM to use other APIs in the cluster to determine those things. This proposal is really about making it possible for clients that may not have rights to read those APIs, or might not know about them, to have some basic information about the cluster.